File: src/test/java/gherkin/pickles/CompilerTest.java
package gherkin.pickles;

import gherkin.AstBuilder;
import gherkin.Parser;
import gherkin.ast.Feature;
import gherkin.deps.com.google.gson.Gson;
import gherkin.deps.com.google.gson.GsonBuilder;
import org.junit.Test;

import java.io.IOException;
import java.util.List;

import static org.junit.Assert.assertEquals;

public class CompilerTest {
    private final Parser<Feature> parser = new Parser<>(new AstBuilder());
    private final Compiler compiler = new Compiler();
    private Gson gson = new GsonBuilder().setPrettyPrinting().create();

    @Test
    public void compiles_a_scenario() throws IOException {
        List<Pickle> pickles = compiler.compile(parser.parse("" +
                "Feature: f\n" +
                "  Scenario: s\n" +
                "    Given passing\n"), "features/hello.feature");


        assertEquals("" +
                        "[\n" +
                        "  {\n" +
                        "    \"name\": \"Scenario: s\",\n" +
                        "    \"steps\": [\n" +
                        "      {\n" +
                        "        \"text\": \"passing\",\n" +
                        "        \"arguments\": [],\n" +
                        "        \"locations\": [\n" +
                        "          {\n" +
                        "            \"path\": \"features/hello.feature\",\n" +
                        "            \"line\": 3,\n" +
                        "            \"column\": 11\n" +
                        "          }\n" +
                        "        ]\n" +
                        "      }\n" +
                        "    ],\n" +
                        "    \"tags\": [],\n" +
                        "    \"locations\": [\n" +
                        "      {\n" +
                        "        \"path\": \"features/hello.feature\",\n" +
                        "        \"line\": 2,\n" +
                        "        \"column\": 3\n" +
                        "      }\n" +
                        "    ]\n" +
                        "  }\n" +
                        "]",
                gson.toJson(pickles));
    }
}

File: test/pickles_test/compiler_test.py
# coding=utf-8
from __future__ import print_function
import json
import textwrap

from gherkin3.token_scanner import TokenScanner
from gherkin3.token_matcher import TokenMatcher
from gherkin3.parser import Parser
from gherkin3.errors import ParserError
from gherkin3.pickles import compiler

from nose.tools import assert_equals, assert_raises


def test_compiles_a_scenario():
    feature_text = textwrap.dedent(
        """\
        Feature: f
          Scenario: s
            Given passing
        """)
    output = Parser().parse(feature_text)
    pickle = compiler.compile(output, 'features/hello.feature')
    expected_pickle = textwrap.dedent(
        """\
        [
          {
            "name": "Scenario: s",
            "steps": [
              {
                "text": "passing",
                "arguments": [],
                "locations": [
                  {
                    "line": 3,
                    "column": 11,
                    "path": "features/hello.feature"
                  }
                ]
              }
            ],
            "tags": [],
            "locations": [
              {
                "line": 2,
                "column": 3,
                "path": "features/hello.feature"
              }
            ]
          }
        ]
        """
    )

    assert_equals(
        pickle,
        json.loads(expected_pickle)
    )


def test_compiles_a_scenario_outline_with_i18n_characters():
    feature_text = textwrap.dedent(
        """\
        Feature: f
          Scenario Outline: with 'é' in title
            Given <with-é>
            Examples:
            | with-é  |
            | passing |
        """)
    output = Parser().parse(feature_text)
    pickle = compiler.compile(output, 'features/hello.feature')
    expected_pickle = textwrap.dedent(
        """\
        [
          {
            "name": "Scenario: with 'é' in title",
            "steps": [
              {
                "text": "passing",
                "arguments": [],
                "locations": [
                  {
                    "line": 6,
                    "column": 5,
                    "path": "features/hello.feature"
                  },
                  {
                    "line": 3,
                    "column": 11,
                    "path": "features/hello.feature"
                  }
                ]
              }
            ],
            "tags": [],
            "locations": [
              {
                "line": 6,
                "column": 5,
                "path": "features/hello.feature"
              },
              {
                "line": 2,
                "column": 3,
                "path": "features/hello.feature"
              }
            ]
          }
        ]
        """
    )

    assert_equals(
        pickle,
        json.loads(expected_pickle)
    )
